packages/@aws-cdk/aws-iotevents/lib/state.ts

```diff
@@ -28,38 +68,93 @@ export class State {
+   * @param options transition options including the condition that causes the state transition
+   */
+  public transitionTo(targetState: State, options: TransitionOptions) {
+    const alreadyAdded = this.transitionEvents.some((event) => event.nextState === targetState);
+    if (alreadyAdded) {
+      throw new Error(`State '${this.stateName}' already has a transition defined to '${targetState.stateName}'`);
+    }
+
+    this.transitionEvents.push({
+      eventName: options.eventName ?? `${this.stateName}_to_${targetState.stateName}`,
+      nextState: targetState,
+      when: options.when,
+    });
+  }
+
+  /**
+   * Collect states in dependency gragh that constructed by state transitions,
+   * and return the JSONs of the states.
+   * This function is called recursively and collect the states.
    *
    * @internal
    */
-  public _toStateJson(): CfnDetectorModel.StateProperty {
-    const { stateName, onEnter } = this.props;
-    return {
-      stateName,
-      onEnter: onEnter && { events: getEventJson(onEnter) },
-    };
+  public _collectStateJsons(collectedStates: Set<State>): CfnDetectorModel.StateProperty[] {
+    if (collectedStates.has(this)) {
+      return [];
+    }
+    collectedStates.add(this);
+
+    return [
+      this.toStateJson(),
+      ...this.transitionEvents.flatMap(transitionEvent => {
+        return transitionEvent.nextState._collectStateJsons(collectedStates);
+      }),
+    ];
   }
 
   /**
-   * Returns true if this state has at least one condition via events.
+   * Returns true if this state has at least one condition as `Event.when`s.
    *
    * @internal
    */
   public _onEnterEventsHaveAtLeastOneCondition(): boolean {
-    return this.props.onEnter?.some(event => event.condition) ?? false;
+    return this.props.onEnter?.some(event => event.when) ?? false;
   }
+
+  private toStateJson(): CfnDetectorModel.StateProperty {
+    const { onEnter } = this.props;
+    return {
+      stateName: this.stateName,
+      onEnter: onEnter && { events: toEventsJson(onEnter) },
+      onInput: {
+        transitionEvents: toTransitionEventsJson(this.transitionEvents),
+      },
+    };
+  }
+}
+
+function toEventsJson(events: Event[]): CfnDetectorModel.EventProperty[] {
+  return events.map(event => {
+    return {
+      eventName: event.eventName,
+      condition: event.when?.evaluate(),
+    };
+  });
 }
 
-function getEventJson(events: Event[]): CfnDetectorModel.EventProperty[] {
-  return events.map(e => {
+function toTransitionEventsJson(transitionEvents: TransitionEvent[]): CfnDetectorModel.TransitionEventProperty[] | undefined {
+  if (transitionEvents.length === 0) {
+    return undefined;
+  }
+
+  return transitionEvents.map(transitionEvent => {
     return {
-      eventName: e.eventName,
-      condition: e.condition?.evaluate(),
+      eventName: transitionEvent.eventName,
+      condition: transitionEvent.when.evaluate(),
+      nextState: transitionEvent.nextState.stateName,
     };
   });
```

Comment:
Like I said, you can shorten this:

```suggestion
  return transitionEvents.map(transitionEvent => ({
    eventName: e.eventName,
    condition: e.condition?.evaluate(),
    eventName: transitionEvent.eventName,
    condition: transitionEvent.when.evaluate(),
    nextState: transitionEvent.nextState.stateName,
  }));
```